### ISSUE TITLE
Issue with devices behind a USB hub

### ISSUE THREAD

**Issue by kaysievers** [adafruit/Adafruit_TinyUSB_ArduinoCore#1]
With commit https://github.com/adafruit/Adafruit_TinyUSB_ArduinoCore/commit/8096c8a8f8aad8458ff5762c990e10ae9dad1a1b the devices seem no longer working when they are connected behind a USB hub. Directly connecting them to the host's root hub works fine.

Reverting the commit makes the device appear.

This is a kernel log from Linux:

```
[34324.981477] usb 1-3: new high-speed USB device number 120 using xhci_hcd
[34325.122916] usb 1-3: New USB device found, idVendor=0bda, idProduct=5411, bcdDevice= 1.21
[34325.122921] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[34325.122924] usb 1-3: Product: 4-Port USB 2.0 Hub
[34325.122927] usb 1-3: Manufacturer: Generic
[34325.126468] hub 1-3:1.0: USB hub found
[34325.128779] hub 1-3:1.0: 4 ports detected
[34325.416491] usb 1-3.1: new full-speed USB device number 121 using xhci_hcd
[34325.492536] usb 1-3.1: device descriptor read/64, error -71
[34325.671522] usb 1-3.1: device descriptor read/64, error -71
[34325.855503] usb 1-3.1: new full-speed USB device number 122 using xhci_hcd
[34325.931543] usb 1-3.1: device descriptor read/64, error -71
[34326.111541] usb 1-3.1: device descriptor read/64, error -71
[34326.219655] usb 1-3-port1: attempt power cycle
[34326.815518] usb 1-3.1: new full-speed USB device number 123 using xhci_hcd
[34326.829286] usb 1-3.1: device descriptor read/8, error -32
[34326.945371] usb 1-3.1: device descriptor read/8, error -32
[34327.128514] usb 1-3.1: new full-speed USB device number 124 using xhci_hcd
[34327.143060] usb 1-3.1: device descriptor read/8, error -32
[34327.265333] usb 1-3.1: device descriptor read/8, error -32
[34327.371687] usb 1-3-port1: unable to enumerate USB device
```

**Comment by hathach** [comment]
Could you tell which example you are running ?

**Comment by kaysievers** [comment]
Good question. Sorry for not testing it properly before reporting this!

I can only reproduce it, and it takes many 'resets', with ```webusb_rgb``` when I remove the
```c
while( !USBDevice.mounted() ) delay(1);
```
line.

Could it be, that there is a race when the sketch immediately tries to prepare an endpoint for incoming packets, while the host is still in progress transmitting the descriptors of the device?

It all looks very timing sensitive. When removing the hub, the host setup might be fast enough not to race against the sketch. Blocking the read() in the sketch until after the USB setup, seems to prevent it too.

If that's the case, what is the proper way to wait in the sketch without blocking in setup()? The sketch should be able to run without USB connected.

**Comment by hathach** [comment]
Don't worry, I am working on a fix, once it is updated, this issue should be resolved.

**Comment by hathach** [comment]
please try to see if commit 8176ab98a61687265bad0048811a2e00b1d796bb works for you

**Comment by kaysievers** [comment]
Yes, nice, it seems there is no difference now, if there is a hub connected in-between or not. Thanks!

One related question, you might have a recommendation:
If the sketch's ```loop()``` calls a USB read() (triggering a ```usbd_edpt_xfer()```) of the device, before the USB setup with the host is done, things get messed up.

This can be prevented by letting the sketch ```setup()``` block. What's the recommended way to handle it, if blocking in ```setup()``` is not an option?

**Comment by hathach** [comment]
which USB read() API you are referring to, tinyusb stack should safely return false for those calls if device is not ready.

**Comment by kaysievers** [comment]
Like here https://github.com/hathach/tinyusb/blob/master/src/class/vendor/vendor_device.c#L96 .

It looks like calling ```tud_vendor_n_read()```too early  would call ```usbd_edpt_xfer()``` during device setup with the host and disturb the communication with the host?

**Comment by hathach** [comment]
@kaysievers yeah, it shouldn't submit an transfer, it is the stack bug, I will update it asap
